### ISSUE TITLE
Why is AvgPool2D taking longer than Conv2D for the same input?

### ISSUE THREAD

**Issue by dhruvbird** [pytorch/pytorch#93188]
### 🐛 Describe the bug

I could be missing something obvious, but when I feed the same input tensor (with shape (20, 1, 28, 28)) to both AvgPool2D and Conv2D, I notice that AvgPool2D takes 56% of the running time and Conv2D takes 39% of the running time. i.e. AvgPool2D is taking 143% of the time that Conv2D is taking for the same input tensor.

Please find the reproduction here: https://colab.research.google.com/drive/1qu7Y9qe6yts7Dp4W4eL8j0Fe-5sgXQvm

Here's the code that can reproduce it (mentioned inline for readability - is the same as in the colab, which also has the output/results).

```
import torch
from torch import nn
from torch.profiler import profile, record_function, ProfilerActivity

with torch.inference_mode():

  avgpool2d = nn.AvgPool2d(3, stride=1)
  conv2d = nn.Conv2d(in_channels=1, out_channels=1, kernel_size=3, padding=0)

  cv2d_input = torch.randn((20, 1, 28, 28))
  ap2d_input = torch.randn((20, 1, 28, 28)) # conv2d(cv2d_input)
  print("Shape of Conv2d Input: {}".format(cv2d_input.shape))
  print("Shape of AvgPool2D Input: {}".format(ap2d_input.shape))

  with profile(activities=[ProfilerActivity.CPU], profile_memory=True, record_shapes=True) as prof:
    for _ in range(1000):
      avgpool2d(ap2d_input)
      conv2d(cv2d_input)

  print(prof.key_averages().table(sort_by="cpu_time_total", row_limit=20))
```

Output:

```
Shape of Conv2d Input: torch.Size([20, 1, 28, 28])
Shape of AvgPool2D Input: torch.Size([20, 1, 28, 28])
----------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                        Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem    # of Calls  
----------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
            aten::avg_pool2d        56.80%     540.816ms        56.80%     540.816ms     540.816us      51.57 Mb      51.57 Mb          1000  
                aten::conv2d         1.11%      10.533ms        42.40%     403.724ms     403.724us      51.57 Mb     211.25 Kb          1000  
           aten::convolution         1.03%       9.773ms        41.54%     395.539ms     395.539us      51.57 Mb       1.03 Mb          1000  
          aten::_convolution         1.09%      10.391ms        41.00%     390.432ms     390.432us      51.57 Mb     211.25 Kb          1000  
    aten::mkldnn_convolution        39.20%     373.237ms        39.98%     380.677ms     380.677us      51.57 Mb       5.93 Mb          1000  
                 aten::empty         0.60%       5.713ms         0.60%       5.713ms       2.857us      51.42 Mb      51.42 Mb          2000  
           aten::as_strided_         0.18%       1.727ms         0.18%       1.727ms       1.727us      -5.78 Mb      -5.78 Mb          1000  
                    [memory]         0.00%       0.000us         0.00%       0.000us       0.000us    -103.15 Mb    -103.15 Mb          2000  
----------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 952.190ms
```

### Versions

1.13.1+cu116

**Comment by sanchitintel** [comment]
convolution is dispatched to oneDNN, which JITs & caches an optimized kernel for the convolution

**Comment by dhruvbird** [comment]
@sanchitintel Make sense! In that case, would it be useful to implement average pooling as a special case of convolution - i.e. if you're averaging K elements, then the kernel can be set up so that each element is 1/K. I checked that conv is faster than average pooling even for a (N x C x H x W) = (20 x 1 x 200 x 200) tensor.

**Comment by mingfeima** [comment]
@dhruvbird  I did not do much optimization on `AvgPool2d` or `MaxPool2d` on channels first memory format (NCHW), the previous focus is channels last (NHWC). But for your case (C=1), channels last should not be better since the kernel did vectorization on C.

Some similar issue are reported, generally `pooling`, `interpolation`, `replication` all have the same issue:
* channels first performance is poor
* channels last (when C< 8) is also poor

Applies to the usage case of raw input data transformation (C=1 for intensity image, C=3 for RGB). Most if the complaints come from pyTorch is slower than NumPy on those usage scenarios. Will have it fixed.

**Comment by sanchitintel** [comment]
> Make sense! In that case, would it be useful to implement average pooling as a special case of convolution - i.e. if you're averaging K elements, then the kernel can be set up so that each element is 1/K. I checked that conv is faster than average pooling even for a (N x C x H x W) = (20 x 1 x 200 x 200) tensor.

Hi @dhruvbird, that sounds like a great idea!

@mingfeima, maybe @dhruvbird's proposal would yield better performance (measured over multiple runs) than an optimized ATen AvgPool kernel, since conv is dispatched to oneDNN? I'll experiment & share perf data later.

**Comment by dhruvbird** [comment]
> > Make sense! In that case, would it be useful to implement average pooling as a special case of convolution - i.e. if you're averaging K elements, then the kernel can be set up so that each element is 1/K. I checked that conv is faster than average pooling even for a (N x C x H x W) = (20 x 1 x 200 x 200) tensor.
> 
> Hi @dhruvbird, that sounds like a great idea!
> 
> @mingfeima, maybe @dhruvbird's proposal would yield better performance (measured over multiple runs) than an optimized ATen AvgPool kernel, since conv is dispatched to oneDNN? I'll experiment & share perf data later.

Thanks @sanchitintel 

If you do get to this please could you also check a multi channel input tensor? In my example, I am using a C=1 image, which may not be representative of most real-world use-cases. The test could probably include a 1 x 32 x 200 x 200 tensor, which is basically a 32 channel tensor image. Also looking at https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html it seems like this would have the same input and output channels, and that convolution would be per-channel, and not across channels, which is something I haven't seen before. https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d suggests that there may be way to do this using groups, but I haven't actually tried it out myself. Maybe there's a trick I'm missing something and am willing to hear what you have to say!

Cheers!

**Comment by mingfeima** [comment]
I would like to offload some oneDNN optimization techniques to pytorch directly to fix this issue rather than using Conv2d for some Pooling layers. Major reason is this would be harder to maintain since Conv2d is a really complexed module. Additionally Conv2d on channels first memory format would trigger reorder which is a burden.

More general request is: improve pooling layer perf on NCHW memory format (together with C=1 and C = 3 NHWC memory format).

The major perf bottleneck on NCHW pooling layers is `index` computation (compute the window for each output index), right now there are many duplicated computations here:
* cache indices for each H and each W
* Vectorization on NCHW is not so straight forward as NHWC. we have to do `gather` on `W`, since the window size are not all the same for each of the output index (borders would have smaller window), the cached indices need masking.

**Comment by jgong5** [comment]
> I would like to offload some oneDNN optimization techniques to pytorch directly to fix this issue rather than using Conv2d for some Pooling layers.

I agree with this direction. In particular, I don't find reasons why we cannot get good performance via direct optimization in PyTorch. In general, when the channel number is big, explicitly converting to channels last is more preferred to get good perf; when the channel number is small, we can assume that the spatial dim is big and we can vectorize on spatial dim with channels first format.

**Comment by dhruvbird** [comment]
> I would like to offload some oneDNN optimization techniques to pytorch directly to fix this issue rather than using Conv2d for some Pooling layers. Major reason is this would be harder to maintain since Conv2d is a really complexed module. Additionally Conv2d on channels first memory format would trigger reorder which is a burden.

> More general request is: improve pooling layer perf on NCHW memory format (together with C=1 and C = 3 NHWC memory format).

I agree that doing it outside of Conv probably makes most sense, since `Conv` needs to handle some additional cases like dilation, etc... which don't show up in average pooling. Average Pooling also has a setting to ignore the padding pixels when computing the average, and it's not clear to me if there's a way to translate that into a Conv operation.

```
count_include_pad [bool] - when True, will include the zero-padding in the averaging calculation
```

**Comment by sanchitintel** [comment]
> Average Pooling also has a setting to ignore the padding pixels when computing the average, and it's not clear to me if there's a way to translate that into a Conv operation.

Thanks again for your inputs, @dhruvbird! I also noticed that a potential _fast-path_ for `AvgPool` that leverages `Conv` can be enabled in very limited cases.

I'll defer to @mingfeima's & @jgong5's proposed solution for resolving this issue.

**Comment by orioninthesky98** [comment]
hi, is there any updates on this issue? 

i was profiling my CNN-like model and I see similar results (but with `avg_pool3d`). if the batch size is really large (32768 or higher) then the `avg_pool3d` takes up more than 80% of the time taken for `model.forward()`, I am wondering what could be the cause
